### PR TITLE
enable coverage

### PR DESCRIPTION
**Pull request by Emilgardis**
Implements coverage via lcov artifacts and uploads them to https://app.codecov.io/gh/cross-rs/cross

### CODE REVIEW

File: .github/actions/cargo-llvm-cov/action.yml [hunk 1]
```diff
@@ -0,0 +1,55 @@
+name: llvm coverage
+description: Setsup everything that is needed for coverage. Makes artifacts available for processing later, prefixed with cov-
```
Comment (Emilgardis) [inline .github/actions/cargo-llvm-cov/action.yml:2]:
```suggestion
description: Sets up everything that is needed for coverage. Makes artifacts available for processing later, prefixed with cov-
```

File: .github/actions/setup-rust/action.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ description: Setup Rust with specified toolchain, target and components.
 inputs:
   toolchain:
     description: 'Toolchain'
-    default: 1.58.1
+    default: 1.60.0
```
Comment (Emilgardis) [inline .github/actions/setup-rust/action.yml:6]: technically, msrv is still 1.58.1, just need 1.60.0 or nightly to use -C coverage

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -322,3 +332,25 @@ jobs:
...
+      - name: Install cargo-llvm-cov
+        uses: taiki-e/install-action@cargo-llvm-cov
+      - uses: actions/download-artifact@v3
+        with:
+          path: ${{ runner.temp }}/artifacts
+      - run: |
+          mkdir profs
+          mv ${artifacts}/*/*.info profs/
+          find profs -name "*.info" -exec echo -a {} \; | xargs lcov -o lcov.info
+
```
Comment (Emilgardis) [inline .github/workflows/ci.yml:354]: here's where we're missing something, need/should report it

**Comment by Emilgardis** [comment]
bors try

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -322,3 +335,39 @@ jobs:
...
+              ) && github.event.head_commit.message) || ''
+            }}
+      - name: Upload to codecov.io
+        run: |
+          set -x
+          curl -Os https://uploader.codecov.io/latest/linux/codecov
+          chmod +x codecov
+          sha_rev=$(git rev-parse HEAD)
+          find ${artifacts} -name "lcov.*.info" -exec ./codecov -F $(echo {} | sed -n 's/lcov\.\(.*\)\.info/\1/p') \
+            ${pr:+-P ${pr}} -f {} --sha ${sha_rev} -n $(echo {} | sed -n 's/lcov\.\(.*\)\.info/\1/p') \;
```
Comment (Emilgardis) [inline .github/workflows/ci.yml:370]: we may have to specify parent sha here, not sure if it really works when you have a staging/trying branch.

been trying to find prior art for codecov with bors integration but not able to find any good examples, only ones I find are my own previous tries :D

File: .github/actions/setup-rust/action.yml (hunk 2 above)
Comment (Alexhuszagh) [reply .github/actions/setup-rust/action.yml:6]: We probably still want to test it compiles against 1.58.1 still, so we might want a separate default version only when checking for code coverage.

File: CHANGELOG.md [hunk 5]
```diff
@@ -73,12 +73,12 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Internal
 
+- Change rust edition to 2021 and bump MSRV for the cross binary to 1.60.0
```
[the remark this answers is not in the document]
Comment (Emilgardis) [reply CHANGELOG.md:76]: this just moves the entry up and changes the rust version

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -321,3 +334,39 @@ jobs:
...
```
Comment (Emilgardis) [inline .github/workflows/ci.yml:369]: I want this to never be able to fail, but on fail echo `::error title=Coverage upload failed::`

Comment (Emilgardis) [reply .github/workflows/ci.yml:369]: I don't know how to do that well though

**Comment by Alexhuszagh** [comment]
Fix the changelog and r=me whenever.

**Review by Alexhuszagh — CHANGES_REQUESTED** [review]
Looking at the merge [attempt](https://github.com/cross-rs/cross/actions/runs/2577693009), we'll likely need to export if the target supports LLVM code coverage. It may be worth just doing it on x86_64-unknown-linux-gnu.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -231,6 +232,18 @@ jobs:
           TARGET: ${{ matrix.target }}
           SUB: ${{ matrix.sub }}
         shell: bash
+
+      - name: LLVM instrument coverage
+        id: cov
+        uses: ./.github/actions/cargo-llvm-cov
+        if: steps.prepare-meta.outputs.has-image
+        with:
```
Comment (Alexhuszagh) [inline .github/workflows/ci.yml:240]: Probably change to:

```yaml
if: steps.prepare-meta.outputs.has-image && matrix.target == 'x86_64-unknown-linux-gnu'
```

**Comment by Emilgardis** [comment]
bors try --target mips64-unknown-linux-muslabi64

File: .github/workflows/ci.yml (hunk 7 above)
Comment (Emilgardis) [reply .github/workflows/ci.yml:240]: We don't need (or want) coverage on the test binaries, fixed this by clearing RUSTFLAGS, maybe it should be more specific though (i.e only remove `instrument-coverage` but not sure how to do that well)

**Review by Alexhuszagh — APPROVED** [review]

**Pull request merged**